src/automate_save_state.py

```diff
@@ -0,0 +1,263 @@
+#!/usr/bin/env python3
+import asyncio
+import subprocess
+import time
+import os
+import sys
+import gzip
+import shutil
+from pathlib import Path
+from playwright.async_api import async_playwright
+
+# Configuration
+PORT = 8888
+VALKEY_VERSION = os.getenv('VALKEY_VERSION', '8.1.0')
+TIMEOUT_SECONDS = 20 * 60  # 20 minutes max wait for boot
+CHECK_INTERVAL = 2  # Check every 2 seconds
+MAX_STATE_SIZE_MB = 40  # Maximum compressed state size in MB
+
+async def wait_for_boot_complete(page, timeout=TIMEOUT_SECONDS):
+    """Wait for boot to complete by checking for server log data"""
+    print("Waiting for boot process to complete...")
+    
+    start_time = time.time()
+    while time.time() - start_time < timeout:
+        try:
+            # Check if server log has content (indicates boot complete)
+            log_content = await page.locator('#log-terminal-container').inner_text()
+            if log_content and len(log_content.strip()) > 0:
+                print("Boot process completed!")
+                return True
+            
+            # Alternative: Check for specific boot completion text
+            # Adjust the selector/text based on actual page behavior
+            await asyncio.sleep(CHECK_INTERVAL)
+            
+        except Exception as e:
+            await asyncio.sleep(CHECK_INTERVAL)
+    
+    print("Timeout waiting for boot to complete")
+    return False
+
+async def test_basic_functionality(page, valkey_version, compressed_state_path):
+    """Test basic functionality of the Valkey image"""
+    print("\nRunning basic functionality tests...")
+    
+    all_tests_passed = True
+    
+    # Test 1: Check if server log contains version
+    print(f"\n1. Testing: Server log contains 'version={valkey_version}'")
+    try:
+        log_content = await page.locator('#log-terminal-container').inner_text()
+        if f'version={valkey_version}' in log_content:
+            print(f"   Version {valkey_version} found in server log")
+        else:
+            print(f"   Version {valkey_version} NOT found in server log")
+            print(f"   Log preview: {log_content[:500]}...")
+            all_tests_passed = False
+    except Exception as e:
+        print(f"   Failed to check server log: {e}")
+        all_tests_passed = False
+    
+    # Test 2: Send PING and check for PONG response
+    print(f"\n2. Testing: PING command responds with PONG")
+    try:
+        # Focus on the terminal
+        terminal = page.locator('#terminal-container')
+        await terminal.click()
+        await asyncio.sleep(1)
+        
+        # Type the PING command
+        await page.keyboard.type('PING\r')
+        print("   Sent: valkey-cli PING")
+        
+        # Wait for response - access xterm.js buffer through serialAdapter
+        pong_found = False
+        for attempt in range(10):  # Try for up to 10 seconds
+            await asyncio.sleep(1)
```

Comment:
seems wierd we choose to always wait first... Maybe we can wait only in case we got an exception?